blog: Project-Scoped Stacks in Self-Managed Backends (#2622)
pulumi/pulumi-hugo: themes/default/content/blog/project-scoped-stacks-in-self-managed-backend/index.md

Reviewer: Hm, this seems to have kinda snuck in under the radar -- and in fact, very little of our writing to date adheres to this style. So I'm not sure I'd call it a recommendation at this point (and in fact I personally strongly prefer the opposite). Let's sync up and see if we can figure out a balance that works for everyone here. 😄

Reviewer 2: Another strong preference for the opposite here :-)

Reviewer 3: Per Slack chat with @cnunciato, adjusted the guide to allow both: https://github.com/pulumi/pulumi-hugo/pull/2633
Regardless of the above PR, I'm also happy to have a longer chat separately about the merits of the style. :)

You no longer have to include the project name in the name of the stack.
You can create a new project and continue to use the short, simple names
that you're used, such as "dev", "staging", "production", etc.
across various projects, without risk of conflict with other projects.

![Project-scoped self-managed stacks](project-scoped-self-managed-stacks.png)

### Upgrading existing backends

As mentioned in the previous section,
only stacks created in new or empty self-managed backends
will be scoped to their projects.
Existing backends will continue to operate as they do today.

If you have an existing self-managed backend that you'd like to upgrade
to use project-scoped stacks,
you can do so with the new `pulumi state upgrade` command,
which upgrades all stacks in the backend to use the new project-scoped layout.

```
$ pulumi state upgrade
This will upgrade the current backend to the latest supported version.
Older versions of Pulumi will not be able to read the new format.
Are you sure you want to proceed?
Please confirm that this is what you'd like to do by typing `yes`:
```

{{% notes type="info" %}}
Once a self-managed backend is upgraded,

> older versions of the Pulumi CLI will be unable to access this backend.

Reviewer 4: I think this should be made clearer, old CLIs can still access the backend they just can't see any of the stack data that's now been moved.

---
title: "Projects in Self-Managed State Backends"
date: 2023-03-29
meta_desc: Pulumi's Self-Managed backend now supports stack names scoped by project name.
meta_image: meta.png
authors:
    - justin-vanpatten
    - abhinav-gupta
tags:
    - features
    - pulumi-releases
---

> At Pulumi, we aim to offer our users the best-in-class IaaC experience.

Reviewer 5: ```suggestion
At Pulumi, we aim to offer our users the best-in-class IaC experience.
```

Reviewer 3: 🤦  Yes. My brain and fingers disagreed about what I was writing.

> We recently shipped a series of impactful enhancements to our Service backend

Reviewer 5: ```suggestion
We recently shipped a series of impactful enhancements to our [Pulumi Service](/product/pulumi-service/) backend
```

> We recently shipped a series of impactful enhancements to our [Pulumi Service](/product/pulumi-service/) backend

Reviewer 5: any other examples we could link to here for recent enhancements?

Reviewer 3: I'm aware of the next big incoming thing for the service (which I didn't want to spoil here), but there weren't too many service-specific announcements recently that I could point to here. 
I figured if I didn't have a list of 3-5 more, I shouldn't say "enhancements like A, B, C, ..., and more on the way", and instead went with suggesting that there are many enhancements—side note, check out deployments, our most recent cool feature.
That said, I haven't worked on the service at all, so I could be wrong about what we want to highlight here?

(check out [Pulumi Deployments](/blog/pulumi-deployments/) if you haven't already!),
and we have more exciting features coming soon.

> While our Service backend remains the best place to store your Pulumi state,

Reviewer 5: instead of "best", perhaps something more descriptive such as "most convenient, reliable, and secure place"

> While our Service backend remains the most ideal place to store your Pulumi state,

Reviewer 5: or highlighting a few features that are valuable

> While our Service backend remains the most ideal place to store your Pulumi state,

Reviewer 3: Maybe "most convenient, reliable, and secure" works here. My initial inclination is that highlighting specifics here takes away from it—"best" is still the only word I can think of that incorporates "this is what everyone should use, it's good, it'll keep getting better, and it'll always be the most desirable choice than alternatives".

Convenience, reliability, and security don't nearly cover the advantages of the service: it just has more features, it has organization support, access control, it handles large company needs, it scales better, it handles multi-user setups better, etc.—it's just better. I don't think we should put all that in this post—this is still a feature announcement for filestate after all.

If you feel strongly that we should use "convenient, reliable, and secure" over "best" (which I still prefer over "ideal", TBH), I'm happy to change it.

> While our Service backend remains the most ideal place to store your Pulumi state,

Reviewer 5: ```suggestion
While we recommend using the Pulumi Service backend to store your state because of its [many features](/blog/six-things-about-pulumi-service/) that increase your productivity,
```

> While our Service backend remains the most ideal place to store your Pulumi state,

Reviewer 5: i just suggested a change. lmk what you think?

> While our Service backend remains the most ideal place to store your Pulumi state,

Reviewer 3: That looks good! Thanks!

Feel free to skip ahead to the next section if you're already familiar with this.

### Projects and stacks

Pulumi [*projects*](/docs/intro/concepts/project/) are a unit of organization.
A project holds the code and configuration for your infrastructure,
and can be used to deploy multiple stacks.
A [*stack*](/docs/intro/concepts/stack/) is a single,
isolated instance of your infrastructure.

For example, you might have an "app" project for your application,
deploying to the "test" and "prod" stacks.

![Projects and stacks](projects-and-stacks.png)

### What's a state backend?

Each stack tracks metadata about your infrastructure
so it can manage your cloud resources.
This metadata is called [*state*](/docs/intro/concepts/state/)
and it is stored in a *backend* of your choosing:
**Service** or **Self-Managed**.

The **Service** backend is hosted at [`app.pulumi.com`](https://app.pulumi.com)
or [on your own server](https://www.pulumi.com/product/self-hosted/).
It provides the best combination of usability, safety, and security for most users.

The **Self-Managed** backend stores state in a JSON file,
deferring to you to manage persistence of this information.

> You can store it in AWS S3, Azure Blob Store, Google Cloud Storage,

Reviewer: FWIW, the actual name of the thing is "Amazon S3". I almost always have to look these things up for myself, as they [tend to follow a pattern](https://stackoverflow.com/questions/33125790/why-some-services-are-called-aws-xxx-and-the-others-amazon-xxx) ... but of course not always. 😄 https://aws.amazon.com/documentation-overview/

```suggestion
You can store it in Amazon S3, Azure Blob Store, Google Cloud Storage,
```

Reviewer 3: Oops, yeah, let me fix this. S3 and VPC are both "Amazon", not "AWS".

---

> title: "Project-Scoped Stacks in Self-Managed State Backends"

Reviewer 6: I don't have an obvious suggestion here, but the title feels too technical to be relatable to most end-users. Even advanced ones. Maybe "Better Stack References in Self-Managed State Backends"? (Because that's what makes your use of the product better.)

Reviewer 3: "Project support in Self-Managed State Backends"?

date: 2023-03-29
meta_desc: The latest Pulumi update includes support for stack names scoped by project name in self-managed state backends.
meta_image: meta.png
authors:
    - justin-vanpatten
tags:
    - features
    - pulumi-releases
---

> The latest Pulumi update includes support for project-scoped stacks in self-managed state backends, making self-managed state backends consistent with how stacks are scoped in the Pulumi Service. Stacks created in new or empty self-managed backends will now be scoped by project by default, and existing backends can be upgraded to be scoped by project. This allows using the same simple stack names such as "dev", "staging", "production", etc. across multiple projects within the same self-managed backend.

Reviewer 7: Yeah... I would start with another paragraph that would frame the existing user pain and mention the issue which has 50+ upvotes.

<!--more-->

## What are State Backends?

> Each Pulumi [stack](/docs/intro/concepts/stack/) has its own metadata about your infrastructure so it can manage your cloud resources. This metadata is called [_state_](/docs/intro/concepts/state/) and it is stored in a _backend_ of your choosing: **Service** or **Self-Managed**.

Reviewer: Just 'cause not a proper noun:

```suggestion
Each Pulumi [stack](/docs/intro/concepts/stack/) has its own metadata about your infrastructure so it can manage your cloud resources. This metadata is called [_state_](/docs/intro/concepts/state/) and it is stored in a _backend_ of your choosing: **Service** or **self-managed**.
```

> The **Service** backend is hosted at <a href="https://app.pulumi.com" target="_blank">`app.pulumi.com`</a> (or self-hosted) and provides the best combination of usability, safety, and security for most users.

Reviewer 8: I would not add this

> The **Service** backend is hosted at <a href="https://app.pulumi.com" target="_blank">`app.pulumi.com`</a> (or self-hosted) and provides the best combination of usability, safety, and security for most users.

Reviewer 6: @jaxxstorm How come?

> **Self-Managed** backends let you manage state yourself, stored as simple JSON files in AWS S3, Azure Blob Store, Google Cloud Storage, an alternatie AWS S3 API compatible server such as Minio or Ceph, or on your local filesystem. Using a self-managed backend trades off some amount of reliability for additional control over where metadata is stored.

Reviewer: ```suggestion
**Self-managed** backends let you manage state yourself, stored as simple JSON files in AWS S3, Azure Blob Store, Google Cloud Storage, an alternative AWS S3 API compatible server such as Minio or Ceph, or on your local filesystem. Using a self-managed backend trades off some amount of reliability for additional control over where metadata is stored.
```

## Upgrading Existing Self-Managed Backends

Existing self-managed backends remain non-scoped until upgraded. You can upgrade an existing self-managed backend using the new `pulumi state upgrade` command. This command will upgrade all stacks in the backend to be scoped by project.

```
$ pulumi state upgrade
This will upgrade the current backend to the latest supported version.
Older versions of Pulumi will not be able to read the new format.
Are you sure you want to proceed?
Please confirm that this is what you'd like to do by typing `yes`:
```

If you've included the project name in your stack names as a way of making the stack name fully unique within a self-managed backend, you can additionally use the `pulumi stack rename` command to clean up your stack names.

## Old Pulumi Versions

Old versions of Pulumi will not be able to see stacks scoped by project. New versions of Pulumi (v3.61.0 or later) will output a warning if an older CLI has created any non-scoped stacks in a self-managed backend that has project-scoped stacks, suggesting to use the `pulumi state upgrade` command to upgrade the stacks.

> ## Referencing Project-Scoped Stacks

Reviewer 3: I agree that this needs a better title, but it's not actually for stack references AFAIK.
This section is more for "how to refer to these new, project-scoped stacks in CLI commands".

@justinvp does this feature also affect stack references?
Or was that just impossible before this?

> ## Referencing Project-Scoped Stacks

Reviewer 4: It was impossible before as all stack refs need to be on the same backend. But I agree it's not quite matching the text of this section. Maybe we should have another section below pointing out this now makes it possible to use just one backend and so use stack refs again.

> Project-scoped stacks in self-managed backends are scoped under a virtual organization, a constant value ("organization"). This allows for a consistent stack identity format across both Service and self-managed backends, e.g. with self-managed backends you can refer to a fully-qualified stack with `organization/my-project/my-stack`, or using the shorthands `organization/my-stack` or `my-stack` (depending on context) just like with the Service backend.

Reviewer 6: Suggest a code example of "this" becomes "this". Single TS example should be sufficient.

> Project-scoped stacks in self-managed backends are scoped under a virtual organization, a constant value ("organization"). This allows for a consistent stack identity format across both Service and self-managed backends, e.g. with self-managed backends you can refer to a fully-qualified stack with `organization/my-project/my-stack`, or using the shorthands `organization/my-stack` or `my-stack` (depending on context) just like with the Service backend.

Reviewer: Breaking this into two sentences might make it a bit easier to take in:

```suggestion
Project-scoped stacks in self-managed backends are scoped under a virtual organization, a constant value ("organization"). This allows for a consistent stack identity format across both Service and self-managed backends. For example, with self-managed backends you can refer to a fully-qualified stack with `organization/my-project/my-stack`, or using the shorthands `organization/my-stack` or `my-stack` (depending on context), just like with the Service backend.
```

> ## Availablity

Reviewer: ```suggestion
## Availability
```

> Support for project-scoped stacks in self-managed backends is now available in Pulumi v3.61.0. Give it a try and share your thoughts with us on [Slack](https://slack.pulumi.com/)!

Reviewer 6: ```suggestion
Support for project-scoped stacks in self-managed backends is now available in Pulumi v3.61.0. Give it a try and share your thoughts with us on [the Pulumi Community Slack](https://slack.pulumi.com/)!
```

Great call to action to close the post!
